Src/WindowsServer/WindowsServer.Shared/AzureWebAppRoleEnvironmentTelemetryInitializer.cs

```diff
@@ -0,0 +1,67 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer
+{
+    using System;
+    using System.Threading;
+
+    using Microsoft.ApplicationInsights.Channel;
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.ApplicationInsights.WindowsServer.Implementation;
+   
+    /// <summary>
+    /// A telemetry initializer that will gather Azure Web App Role Environment context information.
+    /// </summary>
+    public class AzureWebAppRoleEnvironmentTelemetryInitializer : ITelemetryInitializer
+    {
+        /// <summary>Azure Web App name corresponding to the resource name.</summary>
+        private const string WebAppNameEnvironmentVariable = "WEBSITE_SITE_NAME";
+        /// <summary>Azure Web App host that contains site name and slot: sitename-slot.azurewebsites.net.</summary>
+        private const string WebAppHostNameEnvironmentVariable = "WEBSITE_HOSTNAME";
+
+        private string roleInstanceName;
+        private string roleName;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureWebAppRoleEnvironmentTelemetryInitializer" /> class.
+        /// </summary>
+        public AzureWebAppRoleEnvironmentTelemetryInitializer()
+        {
+            WindowsServerEventSource.Log.TelemetryInitializerLoaded(this.GetType().FullName);
+        }
+
+        /// <summary>
+        /// Initializes <see cref="ITelemetry" /> device context.
+        /// </summary>
+        /// <param name="telemetry">The telemetry to initialize.</param>
+        public void Initialize(ITelemetry telemetry)
+        {
+            if (string.IsNullOrEmpty(telemetry.Context.Cloud.RoleName))
+            {
+                var name = LazyInitializer.EnsureInitialized(ref this.roleName, this.GetRoleName);
+                telemetry.Context.Cloud.RoleName = name;
+            }
+
+            if (string.IsNullOrEmpty(telemetry.Context.Cloud.RoleInstance))
+            {
+                var name = LazyInitializer.EnsureInitialized(ref this.roleInstanceName, this.GetRoleInstanceName);
+                telemetry.Context.Cloud.RoleInstance = name;
+            }
+
+            if (string.IsNullOrEmpty(telemetry.Context.GetInternalContext().NodeName))
+            {
+                var name = LazyInitializer.EnsureInitialized(ref this.roleInstanceName, this.GetRoleInstanceName);
```

Comment:
If environment variable doesn't exists will this code attempt to get it again and again? Or it will initialize `this.roleInstanceName` into `string.Empty` and not do it any longer?